proposals/fast-lane-ipns.md

```diff
@@ -0,0 +1,336 @@
+
+##### Proposition
+
+At the high level this proposal promises an efficienecy of centralized publishing and resolution while retaining resilience of distributed system.
+
+Proposed name resolution system is inspired by Domain Name System (DNS) as it introduces athoritative name resolvers _(from now on referred as)_ **name keeper** nodes. But unlike DNS, **name owners** _(private key holders)_ are in charge of choosing a name keepers and retain ability to change a name keeper in the future.
+
+> It is assumed that pinning services would be a natural fit for name keeping as they they already are responsible for keeping the content around and mantain list of user pins.
+
+
+This promises to provide efficiency and speed of centralized systems as peers can cache *name keeper* address(es) for name _(on first resolution)_ and use it in all future resolutions, without disturbing rest of the network.
+
+Name publishing becomes also efficient as name owner can notify *name keeper* directly without disturbing rest of the network.
+
+> Please note that this enables peers to effectively resolve and/or publish name without having to remain online.
+
+###### Initial resolution optimization
+
+To make a very first name resolution nearly as efficient as subsequent ones it is propsoed to develop a new **name routing** service and deploy it on boostrap nodes. Name routing nodes will maintain `IPNS name → multiaddrs[]` mapping enabling peers to resolve an address of a name keeper on a first name resolution. Under assumbiton that reassigment of *name keepers* is going to be rare, overhead on name routing nodes should be insignificant as:
+
+1. Name routers could cache addresses for a long time.
+2. Resolving peers could cache address locally (and require router only first time).
+3. Name routers can deny service to peers resolving name too frequently.
+4. Name routers are not involved in publishing.
+
+It is assumed that name routers will be infrastructure nodes, implying that they can be dialed from limited nodes like browsers _*(Have SSL certificates setup, or have a WebRTC address)*_ and can leverage DHT and PubSub so thay they:
+
+1. Can resolve new name on demand by raced query of DHT and PubSub.
+2. Can become aware of new names through PubSub messages.
+3. Can become aware of name keeper changes through PubSub.
+
+> Note that since name keeper changes are expected to be rare TTL can be really high.
+
+Name routing service does not has to be limited to infrastructure nodes, peer could provide that service and resolve name keeper address from a local cache.
+
+
+###### DNSLink Optimization
+
+Name owners could include addresses to chosen name keepers in the [DNSLink][] record, which would enable name resolution without resolving an address through name routers first.
```

Comment:
How would this interact with IPNS? IPNS doesn't use DNSLink.

Reply:
This is just DNSLink optimization, that would allow you to avoid querying routing nodes described in a previous section. When you resolve IPNS name you'd either already have name keeper address or you query routing nodes to obtain one and cache it forever.

Reply:
IPNS and DNSLink aren't really related. For DNSLink, you just query your local DNS resolver which is almost always very fast.

Reply:
Oh I think I know what the misunderstand is here. Today most DNSLink records are `/ipfs/...` while expectation with this done is that whey will turn into `/ipns/...` and will not require DNS provider specific APIs to do updates.

Does this makes more sense ?

Reply:
That's the first I've heard of that. What's the motivation?

Reply:
> That's the first I've heard of that. What's the motivation?

"What does success look like?" section calls out specifically related goals as success metric. That said motivation specifically is:

1. Browser nodes can not update name (unless....)
2. Not all DNS providers provide programatic API to do record updates. And ones that do have their own custom API so there is no general interface to publishing updates.
3. Usually updating DNS record involves separate authentication, creating yet another hurdle for doing it especially in browsers but also beyond.
4. We could have subscribe style interface for realtime stuff if domain name is just a human readable alias to `/ipns/...`, which seems unrealistic otherwise.

If I'm not mistaken originally https://github.com/protocol/web3-dev-team/pull/42 covered a lot of this before it rescoped.